### DISCUSSION TITLE
TwitterProvider generates wrong authorization URL

### DISCUSSION THREAD

**Discussion by kf106** [nextauthjs/next-auth#2883] [Help] [answered]
### Description 🐜

The TwitterProvider GET_AUTHORIZATION_URL starts with `[object Object]` rather than a URL, and authorization subsequently fails.

```
[next-auth][debug][GET_AUTHORIZATION_URL] {
  url: '[object Object]?oauth_token=XXXXXX&oauth_token_secret=XXXXXXX&oauth_callback_confirmed=true'
}
```

### Is this a bug in your own project?

No

### How to reproduce ☕️

Follow the Twitter code example at https://next-auth.js.org/providers/twitter

### Screenshots / Logs 📽

![twitter-next-auth](https://user-images.githubusercontent.com/25053735/135753695-2ed23034-be0f-4f8f-9399-4f295e55ddbf.png)
+

### Environment 🖥

  System:
    OS: Linux 5.11 Ubuntu 20.04.3 LTS (Focal Fossa)
    CPU: (8) x64 Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
    Memory: 1.82 GB / 15.56 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 14.17.6 - ~/.nvm/versions/node/v14.17.6/bin/node
    npm: 7.22.0 - ~/.nvm/versions/node/v14.17.6/bin/npm
  Browsers:
    Brave Browser: 93.1.29.81
    Chrome: 94.0.4606.54
    Firefox: 92.0
  npmPackages:
    next: ^11.1.2 => 11.1.2 
    next-auth: 4.0.0-beta.3 => 4.0.0-beta.3 
    react: ^17.0.2 => 17.0.2 

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by kf106** [comment]
Note: this is in next-auth 4.0.0-beta.3

**Comment by balazsorban44** [comment]
It has to be a configuration error somewhere, I just tested it and I logged in successfully:

![image](https://user-images.githubusercontent.com/18369201/135754141-a72b8f4e-813d-4a02-8591-204d4ed952f2.png)

Could you paste in your `[...nextauth].js` file?

**Reply by kf106** [reply]
[[...nextauth].js.txt](https://github.com/nextauthjs/next-auth/files/7273341/nextauth.js.txt)

**Comment by kf106** [comment, marked as answer]
If I change line 49 of `node_modules/next-auth/server/lib/oauth/authorization-url.js` to 

```
      const url = `${provider.authorization.url}?${new URLSearchParams({
```

it works. By adding the .url to provider.authorization

**Reply by balazsorban44** [reply]
I see! You probably mean this section:

https://github.com/nextauthjs/next-auth/blob/5149a5d8650b856427a392367b94a60c374a16fc/src/server/lib/oauth/authorization-url.js#L36-L40


The Oauth 1 flow did not receive the updates from #2411, since we plan to deprecate it.

Here is my config:

```ts
    // OAuth 1
    TwitterProvider({
      clientId: process.env.TWITTER_ID,
      clientSecret: process.env.TWITTER_SECRET,
    }),
```

which is working fine. Are you sure you don't try to override the `authorization` param in your code to an object for the Twitter provider?

**Reply by kf106** [reply]
I don't. But I have some confusion as to whether to use `import Twitter from 'next-auth/providers/twitter';` or `import TwitterProvider from 'next-auth/providers/twitter';`

**Reply by kf106** [reply]
Mind you, neither works.

**Reply by balazsorban44** [reply]
Since it's a default import, the name doesn't really matter.

Here is the default Twitter config https://github.com/nextauthjs/next-auth/blob/5149a5d8650b856427a392367b94a60c374a16fc/src/providers/twitter.js#L8

authorization is a string, and not an object. so unless you don't override it, I have to dig deeper

**Reply by kf106** [reply]
Are you looking at 4.0.0-beta.3? I've just `rm -fr node_modules` and run `npm install` and my `package.json` has `"next-auth": "4.0.0-beta.3",` - and - I see the lines for `const url = ...` on line 48.

**Reply by kf106** [reply]
Yes, I have the same string on line 14 of twitter.js

**Reply by kf106** [reply]
But if I put a console.log(provider.authorization) in authorization-url.js just before the const url line, I get:

```
{ url: 'https://api.twitter.com/oauth/authenticate', params: {} }
[next-auth][debug][GET_AUTHORIZATION_URL] {
  url: '[object Object]?...
  ```
  
  So by the time it gets there, it's been changed to an object. I don't even know where I would override it in my code.

**Reply by kf106** [reply]
BTW, LinkedIn, Facebook and Github are working fine.

**Reply by balazsorban44** [reply]
Found it, my bad! https://github.com/nextauthjs/next-auth/pull/2842

introduced a normalization. So internally we can always assume `authorization.url` to be the correct place. I still don't understand how it works for me, but I'll open a PR to fix this. 👀 Thank you for catching!

**Comment by kf106** [comment]
You're welcome. Any ETA on 4.0.0-beta.4 ?

**Reply by balazsorban44** [reply]
Will do a quick release when #2884 has been reviewed since it was a bug that affects all Twitter provider users.

**Reply by balazsorban44** [reply]
Should be released now! https://github.com/nextauthjs/next-auth/releases/tag/v4.0.0-beta.4

**Comment by kf106** [comment]
Thank you, and also thank you for the excellent work you're doing providing this module. I wouldn't want to have to code all of this up from scratch.
